Firestore/core/src/firebase/firestore/model/field_value.cc

```diff
@@ -36,132 +36,395 @@ namespace firebase {
+ public:
+  Type type() const override {
+    return Type::Null;
+  }
 
-size_t ReferenceValue::Hash() const {
-  return util::Hash(key, database_id);
-}
+  std::string ToString() const override {
+    return util::ToString(nullptr);
+  }
 
-FieldValue::FieldValue(const FieldValue& value) {
-  *this = value;
-}
+  ComparisonResult CompareTo(const BaseValue& other) const override {
+    ComparisonResult cmp = CompareTypes(other);
+    if (!util::Same(cmp)) return cmp;
 
-FieldValue::FieldValue(FieldValue&& value) {
-  *this = std::move(value);
-}
+    // Null is only comparable with itself and is defined to be the same.
+    return ComparisonResult::Same;
+  }
+
+  size_t Hash() const override {
+    // std::hash is not defined for nullptr_t.
+    return util::Hash(static_cast<void*>(nullptr));
+  }
+};
+
+template <Type type_enum, typename ValueType>
+class SimpleFieldValue : public FieldValue::BaseValue {
+ public:
+  explicit SimpleFieldValue(ValueType value) : value_(std::move(value)) {
+  }
+
+  Type type() const override {
+    return type_enum;
+  }
+
+  std::string ToString() const override {
+    return util::ToString(value_);
+  }
+
+  ComparisonResult CompareTo(const BaseValue& other) const override {
+    ComparisonResult cmp = CompareTypes(other);
+    if (!util::Same(cmp)) return cmp;
+
+    return Compare(value_, Cast<SimpleFieldValue>(other).value());
+  }
+
+  size_t Hash() const override {
+    return util::Hash(value_);
+  }
 
-FieldValue::~FieldValue() {
-  SwitchTo(Type::Null);
+  const ValueType& value() const {
+    return value_;
+  }
+
+ private:
+  ValueType value_;
+};
+
+class BooleanValue : public SimpleFieldValue<Type::Boolean, bool> {
+ public:
+  using SimpleFieldValue<Type::Boolean, bool>::SimpleFieldValue;
+};
+
+template <Type type_enum, typename ValueType>
+class NumberValue : public SimpleFieldValue<type_enum, ValueType> {
+ public:
+  using SimpleFieldValue<type_enum, ValueType>::SimpleFieldValue;
+
+  ComparisonResult CompareTo(const BaseValue& other) const override;
+};
+
+class IntegerValue : public NumberValue<Type::Integer, int64_t> {
+ public:
+  using NumberValue<Type::Integer, int64_t>::NumberValue;
+};
+
+class DoubleValue : public NumberValue<Type::Double, double> {
+ public:
+  using NumberValue<Type::Double, double>::NumberValue;
+};
+
+template <Type type_enum, typename ValueType>
+ComparisonResult NumberValue<type_enum, ValueType>::CompareTo(
+    const BaseValue& other) const {
+  ComparisonResult cmp = this->CompareTypes(other);
+  if (!util::Same(cmp)) return cmp;
+
+  Type this_type = this->type();
+  Type other_type = other.type();
+
+  if (this_type == Type::Integer) {
+    int64_t this_value = Cast<IntegerValue>(*this).value();
+    if (other_type == Type::Integer) {
+      int64_t other_value = Cast<IntegerValue>(other).value();
+      return Compare(this_value, other_value);
+    } else {
+      double other_value = Cast<DoubleValue>(other).value();
+      return util::ReverseOrder(
```

Comment:
Question: why is `ReverseOrder` necessary here? Consider adding a comment to clarify.

Reply:
It's necessary because `CompareMixedNumber(double, int64_t)` is the only available overload. To reuse this for both cases we reverse the arguments and reverse the result. I've added a comment.

Reply:
Optional: perhaps push this complexity into `CompareMixedNumber`? I.e., provide another overload that does the reversal in its implementation?